File: modules/luci-mod-system/luasrc/model/cbi/admin_system/time.lua
-- Copyright 2008 Steven Barth <steven@midlink.org>
-- Copyright 2011 Jo-Philipp Wich <jow@openwrt.org>
-- Licensed to the public under the Apache License 2.0.

local sys   = require "luci.sys"
local zones = require "luci.sys.zoneinfo"
local fs    = require "nixio.fs"
local conf  = require "luci.config"

local m, s, o
local has_ntpd = fs.access("/usr/sbin/ntpd")

m = Map("system", translate("Time"), translate("Here you can configure the time settings."))
m:chain("luci")


s = m:section(TypedSection, "system", translate("System Time"))
s.anonymous = true
s.addremove = false

o = s:option(DummyValue, "_systime", translate("Local Time"))
o.template = "admin_system/clock_status"

o = s:option(ListValue, "zonename", translate("Timezone"))
o:value("UTC")

--[[
	Change signs for «Etc/GMT» zones.

	The special area of «Etc» is used for some administrative zones, particularly for «Etc/UTC»
	which represents Coordinated Universal Time. In order to conform with the POSIX style,
	those zone names beginning with «Etc/GMT» have their sign reversed from what most people
	expect. In this style, zones west of GMT have a positive sign and those east have
	a negative sign in their name (e.g «Etc/GMT-14» is 14 hours ahead/east of GMT.)
--]]
local TZ_replacements = {
	['Etc/GMT']    = 'GMT',
	['Etc/GMT+1']  = 'GMT-1',
	['Etc/GMT+2']  = 'GMT-2',
	['Etc/GMT+3']  = 'GMT-3',
	['Etc/GMT+4']  = 'GMT-4',
	['Etc/GMT+5']  = 'GMT-5',
	['Etc/GMT+6']  = 'GMT-6',
	['Etc/GMT+7']  = 'GMT-7',
	['Etc/GMT+8']  = 'GMT-8',
	['Etc/GMT+9']  = 'GMT-9',
	['Etc/GMT+10'] = 'GMT-10',
	['Etc/GMT+11'] = 'GMT-11',
	['Etc/GMT+12'] = 'GMT-12',
	['Etc/GMT-1']  = 'GMT+1',
	['Etc/GMT-2']  = 'GMT+2',
	['Etc/GMT-3']  = 'GMT+3',
	['Etc/GMT-4']  = 'GMT+4',
	['Etc/GMT-5']  = 'GMT+5',
	['Etc/GMT-6']  = 'GMT+6',
	['Etc/GMT-7']  = 'GMT+7',
	['Etc/GMT-8']  = 'GMT+8',
	['Etc/GMT-9']  = 'GMT+9',
	['Etc/GMT-10'] = 'GMT+10',
	['Etc/GMT-11'] = 'GMT+11',
	['Etc/GMT-12'] = 'GMT+12',
	['Etc/GMT-13'] = 'GMT+13',
	['Etc/GMT-14'] = 'GMT+14',
}

for i, zone in ipairs(zones.TZ) do
	local zone_exists = fs.access("/usr/share/zoneinfo/" .. zone[1])
	if (zone_exists) then
		local zone_name = TZ_replacements[zone[1]] or zone[1]
		o:value(zone[1], zone_name)
	end
end

function o.write(self, section, value)
	local function lookup_zone(title)
		for _, zone in ipairs(zones.TZ) do
			if zone[1] == title then return zone[2] end
		end
	end

	AbstractValue.write(self, section, value)
	local timezone = lookup_zone(value) or "GMT0"
	self.map.uci:set("system", section, "timezone", timezone)

	if fs.access("/usr/share/zoneinfo/" .. value) then
		fs.unlink("/tmp/localtime")
		fs.symlink("/usr/share/zoneinfo/" .. value, "/tmp/localtime")
	end
end


--
-- NTP
--

if has_ntpd then

	-- timeserver setup was requested, create section and reload page
	if m:formvalue("cbid.system._timeserver._enable") then
		m.uci:section("system", "timeserver", "ntp",
			{
                	server = { "0.openwrt.pool.ntp.org", "1.openwrt.pool.ntp.org", "2.openwrt.pool.ntp.org", "3.openwrt.pool.ntp.org" }
			}
		)

		m.uci:save("system")
		luci.http.redirect(luci.dispatcher.build_url("admin/system", arg[1]))
		return
	end

	local has_section = false
	m.uci:foreach("system", "timeserver", 
		function(s) 
			has_section = true 
			return false
	end)

	if not has_section then

		s = m:section(TypedSection, "timeserver", translate("Time Synchronization"))
		s.anonymous   = true
		s.cfgsections = function() return { "_timeserver" } end

		x = s:option(Button, "_enable")
		x.title      = translate("Time Synchronization is not configured yet.")
		x.inputtitle = translate("Set up Time Synchronization")
		x.inputstyle = "apply"

	else
		
		s = m:section(TypedSection, "timeserver", translate("Time Synchronization"))
		s.anonymous = true
		s.addremove = false

		o = s:option(Flag, "enable", translate("Enable NTP client"))
		o.rmempty = false

		function o.cfgvalue(self)
			return sys.init.enabled("sysntpd")
				and self.enabled or self.disabled
		end

		function o.write(self, section, value)
			if value == self.enabled then
				sys.init.enable("sysntpd")
				sys.call("env -i /etc/init.d/sysntpd start >/dev/null")
			else
				sys.call("env -i /etc/init.d/sysntpd stop >/dev/null")
				sys.init.disable("sysntpd")
			end
		end


		o = s:option(Flag, "enable_server", translate("Provide NTP server"))
		o:depends("enable", "1")


		o = s:option(DynamicList, "server", translate("NTP server candidates"))
		o.datatype = "host(0)"
		o:depends("enable", "1")

		-- retain server list even if disabled
		function o.remove() end

	end
end

return m

File: modules/luci-mod-system/htdocs/luci-static/resources/view/system/time.js
'use strict';
'require ui';
'require uci';
'require rpc';
'require form';
'require tools.time as time';

var callInitList, callInitAction, callSetLocaltime, callGetLocaltime, callTimezone, CBILocalTime;

callInitList = rpc.declare({
	object: 'luci',
	method: 'getInitList',
	params: [ 'name' ],
	expect: { '': {} },
	filter: function(res) {
		for (var k in res)
			return +res[k].enabled;
		return null;
	}
});

callInitAction = rpc.declare({
	object: 'luci',
	method: 'setInitAction',
	params: [ 'name', 'action' ],
	expect: { result: false }
});

callGetLocaltime = rpc.declare({
	object: 'luci',
	method: 'getLocaltime',
	expect: { result: { localtime: 0, localtime_gmtoff: 0 } }
});

callSetLocaltime = rpc.declare({
	object: 'luci',
	method: 'setLocaltime',
	params: [ 'localtime' ],
	expect: { result: { localtime: 0, localtime_gmtoff: 0 } }
});

callTimezone = rpc.declare({
	object: 'luci',
	method: 'getTimezones',
	expect: { '': {} }
});

CBILocalTime = form.DummyValue.extend({
	renderWidget: function(section_id, option_id, cfgvalue) {
		return E([], [
			E('span', {}, [
				E('input', {
					'id': 'localtime',
					'type': 'text',
					'readonly': true,
					'value': time.localtimeToString(cfgvalue)
				})
			]),
			' ',
			E('button', {
				'class': 'cbi-button cbi-button-apply',
				'click': ui.createHandlerFn(this, function(ev) {
					return callSetLocaltime(Math.floor(Date.now() / 1000)).then(
						function(t) {
							ev.target.parentNode.querySelector('#localtime').value =
								time.localtimeToString(t);
						}
					);
				})
			}, _('Sync with browser')),
			' ',
			this.ntpd_support ? E('button', {
				'class': 'cbi-button cbi-button-apply',
				'click': ui.createHandlerFn(this, function(ev) {
					return callInitAction('sysntpd', 'restart').then(
						function() {
							ev.target.parentNode.querySelector('#localtime').value =
								_('Synchronization...');
						}
					);
				})
			}, _('Sync with NTP-Server')) : ''
		]);
	},
});

return L.view.extend({
	load: function() {
		return Promise.all([
			callInitList('sysntpd'),
			callTimezone(),
			callGetLocaltime(),
			uci.load('luci'),
			uci.load('system')
		]);
	},

	render: function(rpc_replies) {
		var ntpd_enabled = rpc_replies[0],
		    timezones = rpc_replies[1],
		    localtime = rpc_replies[2],
		    ntp_setup, ntp_enabled,
		    m, s, o;

		m = new form.Map('system',
			_('Time'),
			_('Here you can configure the time settings.'));

		m.chain('luci');
		m.tabbed = false;

		s = m.section(form.TypedSection, 'system', _('System Time'));
		s.anonymous = true;
		s.addremove = false;

		/*
		 * Time
		 */

		o = s.option(CBILocalTime, '_systime', _('Local Time'));
		o.cfgvalue = function() { return localtime };
		o.ntpd_support = ntpd_enabled;

		o = s.option(form.ListValue, 'zonename', _('Timezone'));
		o.value('UTC');

		var zones = Object.keys(timezones || {}).sort();
		for (var i = 0; i < zones.length; i++)
			o.value(zones[i], timezones[zones[i]].name);

		o.write = function(section_id, formvalue) {
			var tz = timezones[formvalue] ? timezones[formvalue].tzstring : null;
			uci.set('system', section_id, 'zonename', formvalue);
			uci.set('system', section_id, 'timezone', tz);
		};

		/*
		 * NTP
		 */

		if (L.hasSystemFeature('sysntpd')) {
			var default_servers = [
				'0.openwrt.pool.ntp.org', '1.openwrt.pool.ntp.org',
				'2.openwrt.pool.ntp.org', '3.openwrt.pool.ntp.org'
			];

			s = m.section(form.TypedSection, 'timeserver', _("Time Synchronization"))
			s.anonymous = true;
			s.addremove = false;

			o = s.option(form.Flag, 'enabled', _('Enable NTP client'));
			o.rmempty = false;
			o.ucisection = 'ntp';
			o.default = o.disabled;
			o.write = function(section_id, value) {
				ntpd_enabled = +value;

				if (ntpd_enabled && !uci.get('system', 'ntp')) {
					uci.add('system', 'timeserver', 'ntp');
					uci.set('system', 'ntp', 'server', default_servers);
				}

				if (!ntpd_enabled)
					uci.set('system', 'ntp', 'enabled', 0);
				else
					uci.unset('system', 'ntp', 'enabled');

				return callInitAction('sysntpd', 'enable');
			};
			o.load = function(section_id) {
				return (ntpd_enabled == 1 &&
				        uci.get('system', 'ntp') != null &&
				        uci.get('system', 'ntp', 'enabled') != 0) ? '1' : '0';
			};

			o = s.option(form.Flag, 'enable_server', _('Provide NTP server'));
			o.ucisection = 'ntp';
			o.depends('enabled', '1');

			o = s.option(form.Flag, 'use_dhcp', _('Use DHCP advertised servers'));
			o.ucisection = 'ntp';
			o.default = o.enabled;
			o.depends('enabled', '1');

			o = s.option(form.DynamicList, 'server', _('NTP server candidates'));
			o.datatype = 'host(0)';
			o.ucisection = 'ntp';
			o.depends('enabled', '1');
			o.load = function(section_id) {
				return uci.get('system', 'ntp', 'server');
			};
		}

		return m.render().then(function(mapEl) {
			L.Poll.add(function() {
				return callGetLocaltime().then(function(t) {
					mapEl.querySelector('#localtime').value = time.localtimeToString(t);
				});
			});

			return mapEl;
		});
	}
});
